src/tools/languageTools.cpp

```diff
@@ -0,0 +1,73 @@
+#include "tools.h"
+#include "stringTools.h"
+#include <mutex>
+
+namespace kiwix
+{
+
+namespace
+{
+
+// These mappings are not provided by the ICU library, any such mappings can be manually added here
+std::map<std::string, std::string> iso639_3 = {
+    {"atj", "atikamekw"},
+    {"azb", "آذربایجان دیلی"},
+    {"bcl", "central bikol"},
+    {"bgs", "tagabawa"},
+    {"bxr", "буряад хэлэн"},
+    {"cbk", "chavacano"},
+    {"cdo", "閩東語"},
+    {"dag", "Dagbani"},
+    {"diq", "dimli"},
+    {"dty", "डोटेली"},
+    {"eml", "emiliân-rumagnōl"},
+    {"fbs", "српскохрватски"},
+    {"guw", "Gungbe"},
+    {"hbs", "srpskohrvatski"},
+    {"ido", "ido"},
+    {"kbp", "kabɩyɛ"},
+    {"kld", "Gamilaraay"},
+    {"lbe", "лакку маз"},
+    {"lbj", "ལ་དྭགས་སྐད་"},
+    {"map", "Austronesian"},
+    {"mhr", "марий йылме"},
+    {"mnw", "ဘာသာမန်"},
+    {"myn", "mayan"},
+    {"nah", "nahuatl"},
+    {"nai", "north American Indian"},
+    {"nds", "plattdütsch"},
+    {"nrm", "bhasa narom"},
+    {"olo", "livvi"},
+    {"pih", "Pitcairn-Norfolk"},
+    {"pnb", "Western Panjabi"},
+    {"rmr", "Caló"},
+    {"rmy", "romani shib"},
+    {"roa", "romance languages"},
+    {"twi", "twi"},
+};
+
+std::once_flag fillLanguagesFlag;
+
+void fillLanguagesMap()
+{
+    for (auto icuLangPtr = icu::Locale::getISOLanguages(); *icuLangPtr != NULL; ++icuLangPtr) {
+        const kiwix::ICULanguageInfo lang(*icuLangPtr);
+        iso639_3.insert({lang.iso3Code(), lang.selfName()});
+    }
+}
+
+} // unnamed namespace
+
+std::string getLanguageSelfName(const std::string& lang)
+{
+    std::call_once(fillLanguagesFlag, fillLanguagesMap);
+    if (lang == "")
+        return "und";
```

Comment:
It should be better to add this directly in the iso639_3 instead of adding a specific case.

Reply:
Do you think Undetermined is a better name than "und"? It's more meaningful and will look better in kiwix-serve
https://iso639-3.sil.org/code/und

Reply:
Done in the commit